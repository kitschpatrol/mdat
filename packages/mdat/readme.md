<!--+ Warning: Content in HTML comment blocks generated by mdat on 2024-02-08 +-->

<!-- header -->

# mdat

[![NPM Package](https://img.shields.io/npm/v/mdat.svg)](https://npmjs.com/package/mdat)
[![License: MIT](https://img.shields.io/badge/License-MIT-yellow.svg)](https://opensource.org/licenses/MIT)

**CLI tool and library for using comments as content templates in Markdown files.**

<!-- /header -->

> \[!NOTE]\
> **Please see the [Mdat Monorepo readme](http://github.com/kitschpatrol/mdat) for additional context.**

<!-- table-of-contents -->

## Table of contents

- [Overview](#overview)
- [Getting started](#getting-started)
  - [Dependencies](#dependencies)
  - [Installation](#installation)
- [Usage](#usage)
  - [CLI](#cli)
  - [API](#api)
  - [Configuration](#configuration)
- [Creating custom rules](#creating-custom-rules)
- [Maintainers](#maintainers)
- [Acknowledgements](#acknowledgements)
- [Contributing](#contributing)
- [License](#license)

<!-- /table-of-contents -->

## Overview

This is a CLI tool and library implementing the Markdown Autophagic Template (mdat) system, which makes it easy to automate the replacement of placeholder comments in Markdown documents with dynamic content from a variety of sources. `mdat` also validates the structure and content of the Markdown document based on constraints specified in the expansion rules.

A trivial example...

Given placeholder comments in a Markdown file like this:

`some-file.md`

```md
<!-- title -->
```

Run your file through the tool:

```sh
mdat some-file.md
```

To turn it into:

`some-file.md`

```md
<!-- title -->

# mdat

<!-- /title -->
```

In this case, according to a set of rules defined in an external configuration file, `<!-- title -->` was replaced with date from `package.json`. The rule system behind these expansions is simple to define and readily extensible beyond the trivial example above.

The `mdat` CLI and library package sits between the lower-level \[`remark-mdat`] remark plugin, and the higher-level \[`mdat-readme`] package. `mdat` offers flexibility, but if you're just looking to expand some placeholders in your readme files then [`mdat-readme`](../mdat-readme) is the place to start.

## Getting started

### Dependencies

The `mdat` CLI tool requires Node 16+. The exported APIs for expanding Markdown text and documents are ESM-only and share the Node 16+ requirement. `mdat` is implemented in TypeScript and bundles a complete set of type definitions.

### Installation

Install locally to access the CLI commands in a single project or to import the provided APIs:

```sh
npm install remark-mdat
```

Or, install globally for access across your system:

```sh
npm install --global remark-mdat
```

## Usage

> \[!WARNING]\
> **The `mdat` CLI tool directly manipulates the contents of readme files, in close (and perhaps dangerous) proximity to your painstakingly crafted words.**
>
> Please make sure any text you care about is committed before running `mdat`, and never directly modify content inside of the comment expansion blocks.
>
> Set the `--meta` flag on the command to add a warning comment to the top of your file explaining the extra caution demanded around the volatile automated sections of your readme.md.

### CLI

<!-- cli-help -->

#### Command: `mdat`

Note: `expand` is the default and only command at the moment.

This section lists top-level commands for `mdat`.

If no command is provided, `mdat expand` is run by default.

Usage:

```txt
mdat [command] [options]
```

| Command  | Argument                | Description                                                         |
| -------- | ----------------------- | ------------------------------------------------------------------- |
| `expand` | `<files..>` `[options]` | Expand comment placeholders in Markdown files. _(Default command.)_ |

_See the sections below for more information on each subcommand._

#### Subcommand: `mdat expand`

Expand comment placeholders in Markdown files.

Usage:

```txt
mdat expand <files..> [options]
```

| Positional Argument | Description                                                                | Type     |
| ------------------- | -------------------------------------------------------------------------- | -------- |
| `files`             | Markdown file(s) with `mdat` placeholder comments to expand. _(Required.)_ | `string` |

| Option      | Alias | Description                                                                                                                                                                                                   | Type      | Default                                                                       |
| ----------- | ----- | ------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- | --------- | ----------------------------------------------------------------------------- |
| `--config`  |       | Path(s) to files containing mdat configs.                                                                                                                                                                     | `array`   | Configuration is loaded if found from the usual places, or defaults are used. |
| `--rules`   | `-r`  | Path(s) to files containing `mdat` comment expansion rules.                                                                                                                                                   | `array`   |                                                                               |
| `--output`  | `-o`  | Output file directory.                                                                                                                                                                                        | `string`  | Same directory as input file.                                                 |
| `--name`    | `-n`  | Output file name.                                                                                                                                                                                             | `string`  | Same name as input file. Overwrites the input file.                           |
| `--print`   |       | Print the expanded Markdown to stdout instead of saving to a file. Ignores `--output` and `--name` options.                                                                                                   | `boolean` | `false`                                                                       |
| `--prefix`  |       | Require a string prefix before all comments to be considered for expansion. Useful if you have a bunch of non-`mdat` comments in your Markdown file, or if you're willing to trade some verbosity for safety. | `string`  |                                                                               |
| `--meta`    | `-m`  | Embed an extra comment at the top of the generated Markdown noting the date of generation and warning editors that certain sections of the document have been generated dynamically.                          | `boolean` | `false`                                                                       |
| `--check`   | `-c`  | Check the input files for rule violations without expanding them. Identifies things like missing comment placeholders and incorrect placeholder ordering.                                                     | `boolean` | `false`                                                                       |
| `--verbose` |       | Enable verbose logging. All verbose logs and prefixed with their log level and are printed to stderr for ease of redirection.                                                                                 | `boolean` | `false`                                                                       |
| `--help`    | `-h`  | Show help                                                                                                                                                                                                     | `boolean` |                                                                               |
| `--version` | `-v`  | Show version number                                                                                                                                                                                           | `boolean` |                                                                               |

<!-- /cli-help -->

<!-- cli-help-note -->

_Meta note: The entire section above was generated automatically by the [`<!-- cli-help -->`](../mdat-readme/src/lib/rules/cli-help/index.ts) mdat expansion rule provided in `mdat-readme`. It dynamically parses the output from `mdat --help` into a markdown table, recursively calling `--help` on subcommands to build a Markdown representation of the help output._

<!-- /cli-help-note -->

#### Examples

##### Basic

Expand comments in a single Markdown file in-place:

```sh
mdat your-file.md
```

##### Multi-file

Expand comments in multiple Markdown files:

```sh
mdat *.md
```

##### Custom configuration

A number of option flags are exposed on the CLI. Any values set here will override both ambient configuration files and any configuration file referenced passed as options:

```sh
mdat  --prefix 'mm-
```

##### Custom configuration file

```sh
mdat  --config 'custom-config.ts"
```

##### Rule sets

```sh
mdat --rules 'rules.ts' 'more-rules.js' 'yet-more-rules.json'
```

### API

`mdat` exports a collection of functions to abstract the process of expanding placeholder comments into a single call. Type aliases are also provided.

Highlights include:

#### Expand String

```ts
function expandString(markdown: string, config?: ExpandConfig, rules?: ExpandRules): Promise<VFile>
```

Takes a string of Markdown and returns. Note that the returned object is a [VFile](https://github.com/vfile), which includes both the post-conversion Markdown content and additional metadata about the conversion.

To get the Markdown content, simply call `.toString()` on the returned VFile object.

#### Expand File

```ts
function expandFile(
  file: string,
  name?: string,
  output?: string,
  config?: ExpandConfig,
  rules?: ExpandRules,
): Promise<VFile>
```

Similar to `expandString()`, but takes a file path and handles setting an optional destination path and file name.

It's up to the caller to actually save the returned VFile object. The [to-vfile](https://www.npmjs.com/package/to-vfile) library can make this particularly painless:

```ts
import { expandFile } from 'mdat'
import { write } from 'to-vfile'

const file = await expandFiles(...)
await write(file)
```

#### Expand Files

```ts
function expandFiles(
  files: string[],
  name?: string,
  output?: string,
  config?: ExpandConfig,
  rules?: ExpandRules,
): Promise<VFile[]>
```

Like `expandFile()`, but accepts an array of inputs. If an output name is specified, the output files are suffixed with a number to prevent name collisions.

#### Load Config

```ts
function loadConfig<T extends Partial<Options>>(options?: {
  additionalConfig?: Array<T | string> | T | string | undefined
  additionalRules?: Array<Rules | string> | Rules | string | undefined
  configExtensionSchema?: z.ZodObject<any>
  searchFrom?: string
}): Promise<T>
```

This is provided for more advanced use cases. It assists in discovering and loading ambient configuration in your project (e.g. fields in your package.json, or dedicated `mdat` config files). It also dynamically loads, validates, and merges additional `mdat` configuration and rule files into a final `ExpandConfig` object ready to be passed into the [`remark-mdat`](../remark-mdat/readme.md) plugin or one of the API functions like `expandFile()`.

The function is generic to accommodate extensions to the configuration type in derivative projects. ([`mdat-readme`]() uses this approach.)

#### Examples

### Configuration

Expansion rules and certain aspects of global configuration are defined in configuration files which may be discovered automatically by `mdat`, or explicitly provided via command line options or library function arguments as shown above.

`mdat` implements configuration loading via [cosmiconfig](https://github.com/cosmiconfig/cosmiconfig), which means a variety of configuration [locations](https://github.com/cosmiconfig/cosmiconfig?tab=readme-ov-file#searchplaces) and file formats are supported. Configuration may be defined directly in your package.json, or in addition to stand-alone TypeScript files, JavaScript files, YAML, JSON, etc.

TypeScript or JavaScript with JSDoc annotations are recommended for the most flexibility and ease of implementing more advanced rules.

`mdat` also allows arbitrary JSON files to be loaded as rule sets, flattening them so any value may be accessed by using a dot-notation key path as a comment keyword.

#### Configuration file format

The `mdat` configuration file is a record object allowing you to customize aspects of the comment expansion process, and also optionally define expansion rules as well under the `rules` key:

```ts
type MdatConfig = {
  addMetaComment?: boolean
  closingPrefix?: string
  keywordPrefix?: string
  metaCommentIdentifier?: string
  rules?: Rules
}
```

A valid configuration file default-exports an object conforming to the above type.

The configuration file may be located in any location supported by [cosmicconfig](https://github.com/cosmiconfig/cosmiconfig?tab=readme-ov-file#searchplaces). I use a `mdat.config.ts` file in the root of my projects.

#### Rule file format

Rules may also be defined in separate files that default-export a record of rules. The record keys become the keywords used to reference a rule from your comments in Markdown.

```ts
type Rules = Record<string, Rule>

type Rule =
  | string
  | ((options: JsonValue, tree: Root) => Promise<string> | string)
  | Rule[]
  | {
      content: string | Rule[] | ((options: JsonValue, tree: Root) => Promise<string> | string)
      applicationOrder?: number | undefined
      order?: number | undefined
      required?: boolean | undefined
    }
```

This is a bit complex, but it's indented to make defining simple rules simple, while still accommodating more demanding use cases.

Some notes on the type:

- Simple rules can be defined directly on the key, either as strings to replace the comment placeholder, or as sync or async functions returning a string.

- If you need more advanced rules, or wish to define conditions for the validation process, you break the top-level keyword key's value out into an object, where a `content` key on the object is responsible for returning the replacement string, and additional fields are available to define validation constraints.

- Note that `content` can itself take an array of Rule objects, which is useful for creating "compound" rules that combine several others into a single comment keyword.

Since it's a record, multiple rules may be combined in single rules file.

If multiple configuration and rule files are loaded, they are merged. CLI options take precedence over ambient configuration discovered by cosmicconfig, and where multiple configuration or rule files are provided, with the "last" rule of a particular key takes precedence.

## Creating custom rules

See the [Examples section](../remark-mdat/readme.md#examples) of the `remark-mdat` readme, or take a look at the implementation of the [rules bundled with `mdat-readme`](../mdat-readme/src/lib/rules/) for more complex examples.

## Maintainers

[@kitschpatrol](https://github.com/kitschpatrol)

## Acknowledgements

Please see the [monorepo readme](../../readme.md#acknowledgments).

<!-- footer -->

## Contributing

[Issues](https://github.com/kitschpatrol/mdat/issues) and pull requests are welcome.

## License

[MIT](license.txt) © Eric Mika

<!-- /footer -->
