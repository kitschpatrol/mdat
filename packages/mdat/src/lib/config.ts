/* eslint-disable complexity */

import { mdatJsonLoader } from './mdat-json-loader'
import { findPackage } from './utilities'
import chalk from 'chalk'
import { type CosmiconfigResult, cosmiconfig } from 'cosmiconfig'
import fs from 'node:fs/promises'
import path from 'node:path'
import plur from 'plur'
import { type NormalizedPackageJson, readPackage } from 'read-pkg'
import {
	type Options,
	type Rules,
	deepMergeDefined,
	log,
	optionsSchema,
	rulesSchema,
} from 'remark-mdat'
import { type JsonValue } from 'type-fest'
import { z } from 'zod'

export type Config = {
	assetsPath?: string
	packageFile?: string
} & Options

// Reflect defaults and normalization
export type ConfigLoaded = {
	addMetaComment: boolean
	assetsPath: string
	closingPrefix: string
	keywordPrefix: string
	metaCommentIdentifier: string
	packageFile: string | undefined
	rules: Rules
}

const configSchema = optionsSchema
	.merge(
		z.object({
			assetsPath: z.string().optional(),
			packageFile: z.string().optional(),
		}),
	)
	.describe('Config Extension')

/**
 * Generously accept either string paths to .ts, .js, or .json files with
 * `Config` type default exports. Takes a single value, or an array of values which
 * will be merged right to left.
 */
export type ConfigToLoad = Array<Config | string> | Config | string

/**
 * Generously accept either string paths to .ts, .js, or .json files with
 * `Rules` type default exports.
 */
export type RulesToLoad = Array<Rules | string> | Rules | string

// Nomenclature is somewhat obtuse:
// Config:
// File loaded implicitly by cosmiconfig, or passed in as an object or string path to load. Has top-level settings options, plus an optional rules object of its own.
//
// Rules:
// File loaded explicitly by the user, it's a record where each key is a replacement keyword
//
// Rules are merged into a final internal config file

/**
 * Load and validate mdat configuration / rule sets
 * Uses cosmiconfig to search in the usual places.
 * Merge precedence: Additional Config Paths < Search Path < mdat-remark defaults
 *
 * Generic to accommodate additional Config options, so set T to your custom config type if needed. You must provide a matching configExtensionSchema as well.
 */
export async function loadConfig(options?: {
	/**
	 * Additional Config objects to merge.
	 *
	 * Strings are treated as paths to `ts`, `js`, or `json` files with `Config` type default exports. These will be dynamically loaded by Cosmiconfig.
	 * Accepts an individual item, or an array. Objects in the array will be merged right to left.
	 *
	 */
	additionalConfig?: ConfigToLoad
	/**
	 * Additional Rules objects to merge.
	 *
	 * Strings are treated as paths to `ts`, `js`, or `json` files with `Rules` type default exports. These will be dynamically loaded by Cosmiconfig.
	 * Accepts an individual item, or an array. Objects in the array will be merged right to left, and take precedence over any rules in previously loaded Config objects as well.
	 */
	additionalRules?: RulesToLoad
	/** Search for config in specific directories, mainly useful for testing. Cosmiconfig default search paths used if unset. */
	searchFrom?: string
}): Promise<ConfigLoaded> {
	const { additionalConfig, additionalRules, searchFrom } = options ?? {}
	// Set defaults, remark-mdat sets these as well... but it sets them at runtime,
	// we need to set them here so they're on the returned object
	let finalConfig: Config = {
		addMetaComment: false,
		assetsPath: './assets',
		closingPrefix: '/',
		keywordPrefix: '',
		metaCommentIdentifier: '+',
		packageFile: await findPackage(),
		rules: {
			mdat: `Powered by the Markdown Autophagic Template system: [mdat](https://github.com/kitschpatrol/mdat).`,
		},
	}

	// 1. Search and load cosmiconfig locations
	const configExplorer = cosmiconfig('mdat')
	const results = await configExplorer.search(searchFrom)

	if (results) {
		// eslint-disable-next-line @typescript-eslint/no-unsafe-assignment
		const { config, filepath } = results
		log.info(`Using config from "${filepath}"`)
		const configFromObject = getAndValidateConfigFromObject(config, configSchema)
		if (configFromObject) {
			finalConfig = deepMergeDefined(finalConfig, configFromObject)
		}
	}

	// 3. Load and merge additional configs
	if (additionalConfig !== undefined) {
		// Ensure we have an array to work with
		const additionalConfigsArray = Array.isArray(additionalConfig)
			? additionalConfig
			: [additionalConfig]

		for (const configOrPath of additionalConfigsArray) {
			let config: unknown

			if (typeof configOrPath === 'string') {
				// It's probably a path to a config object
				const results = await configExplorer.load(configOrPath)
				if (results === null || results === undefined) continue
				// eslint-disable-next-line @typescript-eslint/no-unsafe-assignment
				const { config: loadedConfig, filepath } = results
				log.info(`Loaded additional config from "${filepath}"`)
				config = loadedConfig
			} else {
				// It's probably a pre-loaded config object
				config = configOrPath
			}

			if (config === undefined) continue

			log.info('Merging configuration object')
			const configFromObject = getAndValidateConfigFromObject(config, configSchema)

			if (configFromObject !== undefined) {
				finalConfig = deepMergeDefined(finalConfig, configFromObject)
			}
		}
	}

	// 4. Load and merge additional rule sets
	if (additionalRules !== undefined) {
		// Ensure we have an array to work with
		const additionalRulesArray = Array.isArray(additionalRules)
			? additionalRules
			: [additionalRules]

		// Create new cosmiconfig loader with custom handling
		// to flatten .json files into rule sets
		const rulesExplorer = cosmiconfig('mdat', {
			loaders: {
				'.json': mdatJsonLoader,
			},
		})

		for (const rulesOrPath of additionalRulesArray) {
			let rules: unknown

			if (typeof rulesOrPath === 'string') {
				// It's a path to a rules object (we hope)
				let results: CosmiconfigResult
				// Special work-around for Cosmiconfig's rather zealous package.json interception
				// So we can treat a package.json as a rule set... but note that this
				// is different from how additionalConfigs is handles, where a _configuration_ object
				// is loaded from package.json, instead of being interpreted as _rules_.
				if (path.basename(rulesOrPath).endsWith('package.json')) {
					const packageJson = await fs.readFile(rulesOrPath, 'utf8')
					const flatJson = mdatJsonLoader(rulesOrPath, packageJson) as JsonValue
					results = {
						config: flatJson,
						filepath: rulesOrPath,
					}
				} else {
					// Let cosmiconfig call the custom loader itself
					results = await rulesExplorer.load(rulesOrPath)
				}

				if (results === null || results === undefined) continue
				// eslint-disable-next-line @typescript-eslint/no-unsafe-assignment
				const { config: loadedRules, filepath } = results
				log.info(`Loaded additional config from "${filepath}"`)
				rules = loadedRules
			} else {
				// It's a pre-loaded rules object
				rules = rulesOrPath
			}

			if (rules === undefined) continue

			log.info('Merging rules into configuration object')
			const configFromRulesObject = getAndValidateConfigFromRulesObject(rules, rulesSchema)

			if (configFromRulesObject !== undefined) {
				finalConfig = deepMergeDefined(finalConfig, configFromRulesObject)
			}
		}
	}

	if (finalConfig.rules) {
		const prettyRules = Object.keys(finalConfig.rules)
			.sort()
			.map((rule) => `"${chalk.bold.green(rule)}"`)
		log.info(
			`Loaded ${chalk.bold(prettyRules.length)} mdat comment expansion ${plur('rule', prettyRules.length)}:`,
		)
		for (const rule of prettyRules) {
			log.info(`\t${rule}`)
		}
	} else {
		log.error('No rules loaded from additional configurations or rules, using default.')
	}

	// Trust that all the non-optional values have been set

	// Set global for rules
	config = finalConfig as ConfigLoaded

	return finalConfig as ConfigLoaded
}

// Helpers

function getAndValidateConfigFromRulesObject(
	rulesObject: unknown,
	rulesSchema: z.ZodSchema,
): Config | undefined {
	if (rulesSchema.safeParse(rulesObject).success) {
		// Wrap the rules object in a config object
		const config = {
			rules: rulesObject as Rules,
		}
		return config as Config
	}

	log.error(
		`Rules object has the wrong shape. Ignoring and using default configuration:\n${JSON.stringify(rulesObject, undefined, 2)}`,
	)
}

function getAndValidateConfigFromObject(
	configObject: unknown,
	configSchema: z.ZodSchema,
): Config | undefined {
	if (configSchema.safeParse(configObject).success) {
		return configObject as Config
	}

	log.error(
		`Config object has the wrong shape. Ignoring and using default configuration:\n${JSON.stringify(configObject, undefined, 2)}`,
	)
}

// Rule helpers

// Caution: Impure functions

// TODO better to pass config into rules?
// Let rules access config for custom behavior
let config: ConfigLoaded | undefined
export async function getConfig(): Promise<ConfigLoaded> {
	// Defaults guarantee all keys will be defined
	// Readme config should be defined at this point when called from one of the
	// API functions that handle config loading, but if not, e.g. in testing,
	// we can load it here

	if (config === undefined) {
		log.warn('getConfig(): config was undefined')
		config ??= await loadConfig()
	}

	return config
}

// Convenience function for rules
// Load as package json only as needed, memoize
// Rules could call this themselves, but this is more convenient and efficient
let packageJson: NormalizedPackageJson | undefined
export async function getPackageJson(): Promise<NormalizedPackageJson> {
	const { packageFile } = await getConfig()
	if (packageFile === undefined) {
		throw new Error('No packageFile found or set in config')
	}

	packageJson ??= await readPackage({ cwd: path.dirname(packageFile) })
	if (packageJson === undefined) {
		throw new Error('No package.json found')
	}

	return packageJson
}
