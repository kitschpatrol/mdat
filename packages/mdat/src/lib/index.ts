export {
	type ExpandFileOptions,
	type ExpandFilesOptions,
	type ExpandStringOptions,
	expandFile,
	expandFiles,
	expandString,
} from './api'
export { loadConfig } from './config'
export {
	type NormalizedRules,
	type Options,
	type Rules,
	getMdatReports,
	log,
	reporterMdat,
} from 'remark-mdat'
