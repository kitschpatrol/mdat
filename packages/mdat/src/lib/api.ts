import { type ConfigToLoad, type RulesToLoad, loadConfig } from './config'
import {
	getCheckProcessor,
	getCleanProcessor,
	getExpandProcessor,
	processFiles,
	processString,
} from './processors'
import { type VFile } from 'vfile'

/**
 * Writing is the responsibility of the caller (e.g. via `await write(result)`)
 */
export async function expandFiles(
	files: string | string[],
	name?: string,
	output?: string,
	config?: ConfigToLoad,
	rules?: RulesToLoad,
): Promise<VFile[]> {
	return processFiles(files, loadConfig, getExpandProcessor, name, output, config, rules)
}

export async function expandString(
	markdown: string,
	config?: ConfigToLoad,
	rules?: RulesToLoad,
): Promise<VFile> {
	return processString(markdown, loadConfig, getExpandProcessor, config, rules)
}

export async function checkFiles(
	files: string | string[],
	name?: string,
	output?: string,
	config?: ConfigToLoad,
	rules?: RulesToLoad,
): Promise<VFile[]> {
	return processFiles(files, loadConfig, getCheckProcessor, name, output, config, rules)
}

export async function checkString(
	markdown: string,
	config?: ConfigToLoad,
	rules?: RulesToLoad,
): Promise<VFile> {
	return processString(markdown, loadConfig, getCheckProcessor, config, rules)
}

export async function collapseFiles(
	files: string | string[],
	name?: string,
	output?: string,
	config?: ConfigToLoad,
	rules?: RulesToLoad,
): Promise<VFile[]> {
	return processFiles(files, loadConfig, getCleanProcessor, name, output, config, rules)
}

export async function collapseString(
	markdown: string,
	config?: ConfigToLoad,
	rules?: RulesToLoad,
): Promise<VFile> {
	return processString(markdown, loadConfig, getCleanProcessor, config, rules)
}
