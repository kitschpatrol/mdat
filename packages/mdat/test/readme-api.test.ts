import { getConfig } from '../src/lib/config'
import { expandReadme, expandReadmeString } from '../src/lib/readme/api'
import fs from 'node:fs/promises'
import { describe, expect, it } from 'vitest'

// Replace matched dates with the placeholder text for stable snapshots
function stripDynamic(text: string): string {
	return text.replaceAll(/\s\d{4}-\d{2}-\d{2}\s/g, ' ****-**-** ')
}

describe('comment expansion', () => {
	it('should expand readme comments', async () => {
		const markdown = await fs.readFile('./test/assets/readme-test.md', 'utf8')
		const result = await expandReadmeString(markdown)
		expect(stripDynamic(result.toString())).toMatchSnapshot()
	})

	it('should expand prefixed comments only', async () => {
		const markdown = await fs.readFile('./test/assets/readme-test.md', 'utf8')
		const result = await expandReadmeString(markdown, {
			keywordPrefix: 'mm-',
		})

		expect(stripDynamic(result.toString())).toMatchSnapshot()
	})

	it('should allow additional rules, and they should override those provided by mdat readme', async () => {
		const markdown = await fs.readFile('./test/assets/readme-test.md', 'utf8')
		const result = await expandReadmeString(markdown, undefined, './test/assets/extra-rules.js')
		expect(stripDynamic(result.toString())).toMatchSnapshot()
	})

	it('should exclude the meta tag if asked', async () => {
		const markdown = await fs.readFile('./test/assets/readme-test.md', 'utf8')
		const result = await expandReadmeString(markdown, {
			addMetaComment: false,
		})
		expect(result.toString()).toMatchSnapshot()
	})

	it('should find the local readme and package.json and expand them', async () => {
		const result = await expandReadme()
		const { packageFile } = await getConfig()
		expect(packageFile).not.toBeUndefined()
		expect(result.toString()).toContain('<!-- /')
	}, 30_000)
})
