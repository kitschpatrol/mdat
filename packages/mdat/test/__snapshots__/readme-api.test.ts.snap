// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`comment expansion > should allow additional rules, and they should override those provided by mdat readme 1`] = `
"<!--+ Warning: Content inside HTML comment blocks was generated by mdat and may be overwritten. +-->

<!-- title {prefix: 'üìù '} -->

# \`extra-rules.js\` has overridden the title

<!-- /title -->

<!-- banner -->

<!-- badges {custom: {Documentation: {image: "https://img.shields.io/badge/-documentation-ffdd00?logo=readthedocs&logoColor=222222", link: "https://kitschpatrol.com/svelte-tweakpane-ui"}}} -->

[![NPM Package mdat](https://img.shields.io/npm/v/mdat.svg)](https://npmjs.com/package/mdat)
[![License: MIT](https://img.shields.io/badge/License-MIT-yellow.svg)](https://opensource.org/licenses/MIT)
[![Documentation](https://img.shields.io/badge/-documentation-ffdd00?logo=readthedocs\\&logoColor=222222)](https://kitschpatrol.com/svelte-tweakpane-ui)

<!-- /badges -->

<!-- short-description -->

**CLI tool and library for using comments as content templates in Markdown files, with helpful presets for readmes.**

<!-- /short-description -->

<!-- table-of-contents -->

## Table of contents

- [Example code](#example-code)
- [Prefix example](#prefix-example)
- [Extra rules example](#extra-rules-example)
- [Contributing](#contributing)
- [License](#license)

<!-- /table-of-contents -->

A long description goes here.

## Example code

The code block below was dynamically embedded by MDAT.

<!-- code {file: "./test/assets/extra-rules.js"} -->

\`\`\`.js
// This is a code sample for testing purposes
// There doesn't seem to be a way to type a default export in JSDoc

/** @satisfies {import('../../src/lib/rules').Rules} */
const rules = {
	extra: '**A bold statement from extra-rules.js**',
	title: {
		applicationOrder: 2,
		content: '# \`extra-rules.js\` has overridden the title',
	},
}

export default rules

\`\`\`

<!-- /code -->

## Prefix example

The comment below only expands if \`--prefix mm-\` was set:

<!-- mm-short-description  -->

## Extra rules example

The comment below only expands if \`--rules extra-rules.js\` was set:

<!-- extra -->

**A bold statement from extra-rules.js**

<!-- /extra -->

<!-- contributing -->

## Contributing

[Issues](https://github.com/kitschpatrol/mdat/issues) and pull requests are welcome.

<!-- /contributing -->

<!-- license -->

## License

[MIT](license.txt) ¬© Eric Mika

<!-- /license -->
"
`;

exports[`comment expansion > should exclude the meta tag if asked 1`] = `
"<!-- title {prefix: 'üìù '} -->

# üìù mdat

<!-- /title -->

<!-- banner -->

<!-- badges {custom: {Documentation: {image: "https://img.shields.io/badge/-documentation-ffdd00?logo=readthedocs&logoColor=222222", link: "https://kitschpatrol.com/svelte-tweakpane-ui"}}} -->

[![NPM Package mdat](https://img.shields.io/npm/v/mdat.svg)](https://npmjs.com/package/mdat)
[![License: MIT](https://img.shields.io/badge/License-MIT-yellow.svg)](https://opensource.org/licenses/MIT)
[![Documentation](https://img.shields.io/badge/-documentation-ffdd00?logo=readthedocs\\&logoColor=222222)](https://kitschpatrol.com/svelte-tweakpane-ui)

<!-- /badges -->

<!-- short-description -->

**CLI tool and library for using comments as content templates in Markdown files, with helpful presets for readmes.**

<!-- /short-description -->

<!-- table-of-contents -->

## Table of contents

- [Example code](#example-code)
- [Prefix example](#prefix-example)
- [Extra rules example](#extra-rules-example)
- [Contributing](#contributing)
- [License](#license)

<!-- /table-of-contents -->

A long description goes here.

## Example code

The code block below was dynamically embedded by MDAT.

<!-- code {file: "./test/assets/extra-rules.js"} -->

\`\`\`.js
// This is a code sample for testing purposes
// There doesn't seem to be a way to type a default export in JSDoc

/** @satisfies {import('../../src/lib/rules').Rules} */
const rules = {
	extra: '**A bold statement from extra-rules.js**',
	title: {
		applicationOrder: 2,
		content: '# \`extra-rules.js\` has overridden the title',
	},
}

export default rules

\`\`\`

<!-- /code -->

## Prefix example

The comment below only expands if \`--prefix mm-\` was set:

<!-- mm-short-description  -->

## Extra rules example

The comment below only expands if \`--rules extra-rules.js\` was set:

<!-- extra -->

<!-- contributing -->

## Contributing

[Issues](https://github.com/kitschpatrol/mdat/issues) and pull requests are welcome.

<!-- /contributing -->

<!-- license -->

## License

[MIT](license.txt) ¬© Eric Mika

<!-- /license -->
"
`;

exports[`comment expansion > should expand prefixed comments only 1`] = `
"<!--+ Warning: Content inside HTML comment blocks was generated by mdat and may be overwritten. +-->

<!-- title {prefix: 'üìù '} -->

<!-- banner -->

<!-- badges {custom: {Documentation: {image: "https://img.shields.io/badge/-documentation-ffdd00?logo=readthedocs&logoColor=222222", link: "https://kitschpatrol.com/svelte-tweakpane-ui"}}} -->

<!-- short-description -->

<!-- table-of-contents -->

A long description goes here.

## Example code

The code block below was dynamically embedded by MDAT.

<!-- code {file: "./test/assets/extra-rules.js"} -->

## Prefix example

The comment below only expands if \`--prefix mm-\` was set:

<!-- mm-short-description  -->

**CLI tool and library for using comments as content templates in Markdown files, with helpful presets for readmes.**

<!-- /mm-short-description -->

## Extra rules example

The comment below only expands if \`--rules extra-rules.js\` was set:

<!-- extra -->

<!-- contributing -->

<!-- license -->
"
`;

exports[`comment expansion > should expand readme comments 1`] = `
"<!--+ Warning: Content inside HTML comment blocks was generated by mdat and may be overwritten. +-->

<!-- title {prefix: 'üìù '} -->

# üìù mdat

<!-- /title -->

<!-- banner -->

<!-- badges {custom: {Documentation: {image: "https://img.shields.io/badge/-documentation-ffdd00?logo=readthedocs&logoColor=222222", link: "https://kitschpatrol.com/svelte-tweakpane-ui"}}} -->

[![NPM Package mdat](https://img.shields.io/npm/v/mdat.svg)](https://npmjs.com/package/mdat)
[![License: MIT](https://img.shields.io/badge/License-MIT-yellow.svg)](https://opensource.org/licenses/MIT)
[![Documentation](https://img.shields.io/badge/-documentation-ffdd00?logo=readthedocs\\&logoColor=222222)](https://kitschpatrol.com/svelte-tweakpane-ui)

<!-- /badges -->

<!-- short-description -->

**CLI tool and library for using comments as content templates in Markdown files, with helpful presets for readmes.**

<!-- /short-description -->

<!-- table-of-contents -->

## Table of contents

- [Example code](#example-code)
- [Prefix example](#prefix-example)
- [Extra rules example](#extra-rules-example)
- [Contributing](#contributing)
- [License](#license)

<!-- /table-of-contents -->

A long description goes here.

## Example code

The code block below was dynamically embedded by MDAT.

<!-- code {file: "./test/assets/extra-rules.js"} -->

\`\`\`.js
// This is a code sample for testing purposes
// There doesn't seem to be a way to type a default export in JSDoc

/** @satisfies {import('../../src/lib/rules').Rules} */
const rules = {
	extra: '**A bold statement from extra-rules.js**',
	title: {
		applicationOrder: 2,
		content: '# \`extra-rules.js\` has overridden the title',
	},
}

export default rules

\`\`\`

<!-- /code -->

## Prefix example

The comment below only expands if \`--prefix mm-\` was set:

<!-- mm-short-description  -->

## Extra rules example

The comment below only expands if \`--rules extra-rules.js\` was set:

<!-- extra -->

<!-- contributing -->

## Contributing

[Issues](https://github.com/kitschpatrol/mdat/issues) and pull requests are welcome.

<!-- /contributing -->

<!-- license -->

## License

[MIT](license.txt) ¬© Eric Mika

<!-- /license -->
"
`;
